File: python/cache/persistence.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, Optional

from .storage import Store


class Persistence:
    def __init__(self, snapshot_path: str, log_path: str) -> None:
        self._snapshot_path = Path(snapshot_path)
        self._log_path = Path(log_path)

    def append_set(self, key: str, value: Any, expires_at: Optional[int], version: int) -> None:
        self._append(
            {
                "op": "set",
                "key": key,
                "value": value,
                "expires_at": expires_at,
                "version": version,
            }
        )

    def append_delete(self, key: str, version: int) -> None:
        self._append({"op": "del", "key": key, "version": version})

    def append_expire(self, key: str, expires_at: int, version: int) -> None:
        self._append(
            {"op": "expire", "key": key, "expires_at": expires_at, "version": version}
        )

    def snapshot(self, store: Store) -> None:
        data: Dict[str, Dict[str, Any]] = {}
        for key, entry in store.items():
            if entry.expires_at is not None and store._clock() >= entry.expires_at:
                continue
            data[key] = {
                "value": entry.value,
                "expires_at": entry.expires_at,
                "version": entry.version,
            }
        self._snapshot_path.parent.mkdir(parents=True, exist_ok=True)
        self._snapshot_path.write_text(json.dumps(data))
        if self._log_path.exists():
            self._log_path.write_text("")

    def load(self, store: Store) -> None:
        if self._snapshot_path.exists():
            raw = self._snapshot_path.read_text()
            if raw.strip():
                snapshot = json.loads(raw)
                self._apply_snapshot(store, snapshot)
        if self._log_path.exists():
            for line in self._log_path.read_text().splitlines():
                if line.strip():
                    self._apply_log_entry(store, json.loads(line))

    def _append(self, record: Dict[str, Any]) -> None:
        self._log_path.parent.mkdir(parents=True, exist_ok=True)
        with self._log_path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(record))
            handle.write("\n")

    def _apply_snapshot(self, store: Store, snapshot: Dict[str, Any]) -> None:
        now = store._clock()
        for key, data in snapshot.items():
            expires_at = data.get("expires_at")
            if expires_at is not None and now >= expires_at:
                continue
            ttl_ms = None if expires_at is None else max(expires_at - now, 0)
            version = int(data.get("version", 0))
            existing = store.get_entry(key)
            if existing and version <= existing.version:
                continue
            store.set(key, data.get("value"), ttl_ms=ttl_ms, version=version)

    def _apply_log_entry(self, store: Store, record: Dict[str, Any]) -> None:
        op = record.get("op")
        key = record.get("key")
        version = int(record.get("version", 0))
        existing = store.get_entry(key) if key else None
        if existing and version <= existing.version:
            return
        if op == "set":
            expires_at = record.get("expires_at")
            ttl_ms = None
            if expires_at is not None:
                ttl_ms = max(expires_at - store._clock(), 0)
            store.set(key, record.get("value"), ttl_ms=ttl_ms, version=version)
        elif op == "del":
            store.delete(key)
        elif op == "expire":
            expires_at = record.get("expires_at")
            if expires_at is None:
                return
            ttl_ms = max(expires_at - store._clock(), 0)
            store.expire(key, ttl_ms, version=version)

File: go/cache/persistence.go
package cache

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
)

type Persistence struct {
	snapshotPath string
	logPath      string
	mu           sync.Mutex
}

type snapshotEntry struct {
	Value     any    `json:"value"`
	ExpiresAt *int64 `json:"expires_at"`
	Version   uint64 `json:"version"`
}

type logRecord struct {
	Op        string `json:"op"`
	Key       string `json:"key"`
	Value     any    `json:"value,omitempty"`
	ExpiresAt *int64 `json:"expires_at,omitempty"`
	Version   uint64 `json:"version"`
}

func NewPersistence(snapshotPath, logPath string) *Persistence {
	return &Persistence{snapshotPath: snapshotPath, logPath: logPath}
}

func (p *Persistence) AppendSet(key string, value any, expiresAt *int64, version uint64) {
	p.append(logRecord{Op: "set", Key: key, Value: value, ExpiresAt: expiresAt, Version: version})
}

func (p *Persistence) AppendDelete(key string, version uint64) {
	p.append(logRecord{Op: "del", Key: key, Version: version})
}

func (p *Persistence) AppendExpire(key string, expiresAt int64, version uint64) {
	p.append(logRecord{Op: "expire", Key: key, ExpiresAt: &expiresAt, Version: version})
}

func (p *Persistence) Snapshot(store *Store) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	items := store.Items()
	data := map[string]snapshotEntry{}
	for key, entry := range items {
		var expiresAt *int64
		if entry.HasExpiry {
			exp := entry.ExpiresAt
			expiresAt = &exp
		}
		data[key] = snapshotEntry{
			Value:     entry.Value,
			ExpiresAt: expiresAt,
			Version:   entry.Version,
		}
	}
	if err := os.MkdirAll(filepath.Dir(p.snapshotPath), 0o755); err != nil {
		return err
	}
	raw, err := json.Marshal(data)
	if err != nil {
		return err
	}
	if err := os.WriteFile(p.snapshotPath, raw, 0o644); err != nil {
		return err
	}
	_ = os.WriteFile(p.logPath, []byte{}, 0o644)
	return nil
}

func (p *Persistence) Load(store *Store) error {
	if _, err := os.Stat(p.snapshotPath); err == nil {
		raw, err := os.ReadFile(p.snapshotPath)
		if err != nil {
			return err
		}
		if len(raw) > 0 {
			var snapshot map[string]snapshotEntry
			if err := json.Unmarshal(raw, &snapshot); err != nil {
				return err
			}
			p.applySnapshot(store, snapshot)
		}
	}
	if _, err := os.Stat(p.logPath); err == nil {
		file, err := os.Open(p.logPath)
		if err != nil {
			return err
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			line := scanner.Bytes()
			if len(line) == 0 {
				continue
			}
			var record logRecord
			if err := json.Unmarshal(line, &record); err != nil {
				return err
			}
			p.applyLogRecord(store, record)
		}
		if err := scanner.Err(); err != nil {
			return err
		}
	}
	return nil
}

func (p *Persistence) append(record logRecord) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if err := os.MkdirAll(filepath.Dir(p.logPath), 0o755); err != nil {
		return
	}
	file, err := os.OpenFile(p.logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	raw, err := json.Marshal(record)
	if err != nil {
		return
	}
	writer.Write(raw)
	writer.Write([]byte("\n"))
	writer.Flush()
}

func (p *Persistence) applySnapshot(store *Store, snapshot map[string]snapshotEntry) {
	now := store.Now()
	for key, entry := range snapshot {
		if entry.ExpiresAt != nil && now >= *entry.ExpiresAt {
			continue
		}
		ttlMs := (*int64)(nil)
		if entry.ExpiresAt != nil {
			value := *entry.ExpiresAt - now
			ttlMs = &value
		}
		if existing, ok := store.GetEntry(key); ok && entry.Version <= existing.Version {
			continue
		}
		version := entry.Version
		store.Set(key, entry.Value, ttlMs, &version)
	}
}

func (p *Persistence) applyLogRecord(store *Store, record logRecord) {
	if record.Key == "" {
		return
	}
	if existing, ok := store.GetEntry(record.Key); ok && record.Version <= existing.Version {
		return
	}
	switch record.Op {
	case "set":
		ttlMs := (*int64)(nil)
		if record.ExpiresAt != nil {
			value := *record.ExpiresAt - store.Now()
			ttlMs = &value
		}
		version := record.Version
		store.Set(record.Key, record.Value, ttlMs, &version)
	case "del":
		store.Delete(record.Key)
	case "expire":
		if record.ExpiresAt == nil {
			return
		}
		ttlMs := *record.ExpiresAt - store.Now()
		version := record.Version
		store.Expire(record.Key, ttlMs, &version)
	}
}
